extension/bearertokenauthextension/bearertokenauth.go

```diff
@@ -171,6 +174,38 @@ func (b *BearerTokenAuth) RoundTripper(base http.RoundTripper) (http.RoundTrippe
 	}, nil
 }
 
+// Authenticate checks whether the given context contains valid auth data.
+func (b *BearerTokenAuth) Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error) {
+	metadata := client.NewMetadata(headers)
+	authHeaders := metadata.Get("authorization")
+	if len(authHeaders) == 0 {
+		return ctx, errors.New("authentication didn't succeed")
+	}
+
+	// we only use the first header, if multiple values exist
+	parts := strings.Split(authHeaders[0], " ")
+
+	errInvalidAuthorization := errors.New("invalid authorization header format")
+	var token string
```

Comment:
I am not sure I understand the logic here, would you mind explaining it to me?

Reply:
Sure, but i restructured the logic a bit and added a comment. Let me know if that helps.
```go
// Read the token from the authHeader parts. If a scheme exists, the second entry is used.
// If no scheme exists, we assume the token is in the first entry.
// Example:
// If the authorization header is "Bearer randomtoken", the parts list will be
// ["Bearer", "randomtoken"].
// The scheme "Bearer" will be ignored, and the second entry "randomtoken" will be validated.
// If the authorization header is "randomtoken", the parts list will be ["randomtoken"].
// In this case, the first entry "randomtoken" will be validated.
```